### DISCUSSION TITLE
Support deno?

### DISCUSSION THREAD

**Discussion by tom-sherman** [statelyai/xstate#1625] [Ideas]
It would be great if xstate could support deno out of the box. `core` should be a relatively easy starting point because of it's lack of dependencies.

What do you think?

**Comment by Andarist** [comment]
What does it mean to support deno?

**Reply by tom-sherman** [reply]
I was thinking a package that's published to [denoland](https://deno.land/x) so that it can be easily imported. I believe right now you need to install xstate via npm use https://deno.land/std@0.77.0/node to import it.

**Reply by davidkpiano** [reply]
@tom-sherman I would appreciate a PR for this, if you know what to do. I haven't used Deno yet.

**Reply by Andarist** [reply]
So deno has been advertising no package manager approach and yet they somewhat have one now (although maybe its limited, i dont know the details) in a form of deno.land?

**Reply by Andarist** [reply]
From what i understand deno.land is using GitHub as the source of truth which means that it requires for consumable files to actually be committed. This worries me quite a bit because we want to have a freedom regarding build-time processing of source files and we certainly dont want to commit dist files with each change (its noisy as hell and produces a lot of merge conflicts over time)

**Reply by davidkpiano** [reply]
I have added the webhooks, hopefully that's sufficient: https://deno.land/x/xstate

**Reply by tomByrer** [reply]
@ link:
> No uploaded versions
> This module name has been reserved for a repository, but no versions have been uploaded yet. Modules that do not upload a version within 30 days of registration will be removed. 

Might have to make another release?

**Reply by tomByrer** [reply]
@tom-sherman is XState working in your Dino project now?  At least the dino_land link is not 404 now...

**Reply by tom-sherman** [reply]
Hmmmm, this doesn't seem right: https://deno.land/x/xstate@@xstate/react@1.0.3/react@1.0.3/react@1.0.3

I can't see what module to import and requesting that with curl gives me a maximum redirect limit.

**Reply by tomByrer** [reply]
@davidkpiano

**Reply by Andarist** [reply]
@tomByrer please do not ping maintainers like this. We literally try to read everything that happens in this repository and you have pinged David after only 5h from the latest post in this thread.

> I can't see what module to import and requesting that with curl gives me a maximum redirect limit.

This seems like a denoland issue - `@` is a valid git tag character and it seems to be tripping when encountering it.

**Reply by tom-sherman** [reply]
Oh I see, I wasn't aware that `react@1.0.3` is a tag. It's definitely plausible that it's a denoland issue, could also be an issue with webhook setup?

**Reply by Andarist** [reply]
It's unlikely - the webhook only contains the repository (?) name and the subdirectory. I'm therefore also not sure if this is supposed to work with monorepos.

**Reply by tomByrer** [reply]
https://deno.land/x/xstate is `404 - Not Found` :(

I am starting to use deno for local file running.

**Comment by tomByrer** [comment]
@tom-sherman  have you tried another service, like [jsDelivr](https://www.jsdelivr.com/) or unpkg?

**Reply by tom-sherman** [reply]
Using unpkg I get an error from environment.js, it's trying to access `process.env` so assuming it's in a node environment. This would need to be fixed I guess.

Also it's a little annoying having to specify the declaration every time, but I can live with it

```ts
// @deno-types="https://unpkg.com/xstate@4.14.0/es/index.d.ts"
import {Machine} from 'https://unpkg.com/xstate@4.14.0/es/index.js'
```

**Reply by Andarist** [reply]
> Using unpkg I get an error from environment.js, it's trying to access process.env so assuming it's in a node environment. This would need to be fixed I guess.

I would assume that to support this we need to provide a special deno build that would remove `process.env`. Most likely we'd need to decide what would be a default for this though - would it be a dev or a production build? Dev builds have some nice extra validations, helpful warnings etc but this inquire some extra runtime costs.

> Also it's a little annoying having to specify the declaration every time, but I can live with it

What's an exact effect of this on the Deno program?

---

Is there any comprehensive guide on how one should try to add support for Deno to their project? It's a little bit blurry to me what we should do exactly and I feel like there should be some guidance around this somewhere - otherwise, this feels like a lot of guessing on our side since we are not using Deno ourselves.

**Reply by tom-sherman** [reply]
> What's an exact effect of this on the Deno program?

You need the `@deno-types` comment to get proper completion/type checking if you're importing a JS file, Deno doesn't automatically resolve TS declarations. This is why "deno native" modules will be distributed in TypeScript.

**Reply by Andarist** [reply]
So the actual ideal solution would be to distribute `.ts` files for Deno's consumption? That way you wouldn't have to provide this comment, right? Isn't this a very annoying DX issue that it doesn't try to resolve matching TS declaration files? To understand the overall story better - you are always encouraged to actually build your project using Deno so the production runtime won't have to compile TS, right?

**Reply by tom-sherman** [reply]
> So the actual ideal solution would be to distribute .ts files for Deno's consumption?

Exactly, here's an example of a project that is distributed directly for deno: https://deno.land/x/oak@v6.3.2 I'm sure there also exist some projects that are targeting both npm and deno.land too.

> Isn't this a very annoying DX issue that it doesn't try to resolve matching TS declaration files? 

A little annoying, but deno projects have a convention of a `deps.ts` file where you import all of your external dependencies and re-export them so it's not too big of a deal. https://deno.land/manual/linking_to_external_code#it-seems-unwieldy-to-import-urls-everywhere

> you are always encouraged to actually build your project using Deno so the production runtime won't have to compile TS, right?

I am not too sure on this, but I don't think that is a hard and fast rule. You can `deno bundle main.ts` your application into a single JS file, but it's not required. You can just as easily perform `deno cache main.ts` and then `deno run main.ts`.

**Reply by Andarist** [reply]
Ok, this gives me a better picture of how Deno expects this to be handled. I will think about what it would take for us to support this better. The main challenge is that we need to customize our build & release pipeline to take this into account - this shouldn't be hard but is also not super trivial and I have some more pressing issues to do right now, so this is not on the top of my priority list. Unless you guys would like to help us with this - I could provide guidance for the required changes (most likely those are needed both here and in the [Changesets](https://github.com/atlassian/changesets)).

**Reply by tom-sherman** [reply]
I'd love to work on it but as you mention, it's difficult to find the time! I think one of the more difficult aspects is compiling the imports to be to `.ts`, there is some discussion about that here: https://github.com/denoland/deno/issues/3196

**Reply by tomByrer** [reply]
Thanks Tom!  Seems the better solutions start at this comment, but there are more following:

https://github.com/denoland/deno/issues/3196#issuecomment-662179971

**Reply by tom-sherman** [reply]
Interesting solution! I've also been thinking about a babel plugin for just compiling the imports, this would be helpful for dependency-less libraries like xstate to build a Deno-compatible library.

**Reply by tomByrer** [reply]
If there are ESM packages in the repo, then seems one can use esm.run to use that code in Deno.

https://www.jsdelivr.com/esm

**Reply by virtuoushub** [reply]
FWIW I got this to work:

```typescript
// xstate-deno-demo.ts
import { Machine, interpret } from "https://cdn.skypack.dev/xstate";

// Stateless machine definition
// machine.transition(...) is a pure function used by the interpreter.
const toggleMachine = Machine({
  id: "toggle",
  initial: "inactive",
  states: {
    inactive: { on: { TOGGLE: "active" } },
    active: { on: { TOGGLE: "inactive" } },
  },
});

// Machine instance with internal state
const toggleService = interpret(toggleMachine)
  .onTransition((state: any) => console.log(state.value))
  .start();
// => 'inactive'

toggleService.send("TOGGLE");
// => 'active'

toggleService.send("TOGGLE");
// => 'inactive'
```

```shell
$ deno run xstate-deno-demo.ts 
inactive
active
inactive
```

[edit] add deno version info:

```shell
deno --version
deno 1.6.0 (release, x86_64-apple-darwin)
v8 8.8.278.2
typescript 4.1.2
```

**Reply by tomByrer** [reply]
3rd try: Works for me also, forgot `choose` is imported via `actions.choose`

https://jsfiddle.net/tomByrer/sg7tkwqb/9/

Deno runtime online:
https://www.mycompiler.io/view/B8EgR64

**Comment by tomByrer** [comment]
Should this be moved into Issues to track as a bug-request, or is deno out-of-scope?

**Reply by davidkpiano** [reply]
It's a nice-to-have; let's keep it here (under "Ideas"). I'd gladly accept a PR for someone to work on this, because I haven't used Deno yet.

**Comment by tom-sherman** [comment]
Update: With Deno 1.25's npm support you can run xstate (I've only tried fsm right now but I'm sure the xstate package is the same) directly from npm in Deno:

```js
// script.js
import { createMachine } from "npm:@xstate/fsm";

const machine = createMachine({
  initial: "idle",
  states: {
    idle: {
      on: {
        foo: "done",
      },
    },
    done: {
      type: "final",
    },
  },
});

console.log(machine.transition("idle", "foo").value); // "done"
```

```
> deno run --unstable --allow-env script.js
done
```

🙌 

Few issues right now:

- It's unstable
- It requires permissions to read `process.env.NODE_ENV`. Bit of a bummer but I think there has been some discussions to remove this from the library?
- I have no idea how to get type definitions to load in vs code

The first two problems make it unsuitable to use in libraries right now (my current use case) and the third problem makes it unsuitable to use in your deno app if you're using TypeScript.

Good progress though!

**Reply by Andarist** [reply]
> It requires permissions to read process.env.NODE_ENV. Bit of a bummer but I think there has been some discussions to remove this from the library?

It will be removed in the next major version.

**Reply by tom-sherman** [reply]
I honestly don't think XState needs to do anything apart from that except wait for Deno's support to progress and mature.

**Comment by kindkitchen** [comment]
#### I use import by this url:
#### [https://esm.sh/xstate@5.6.0/dist/xstate.esm.js](https://esm.sh/xstate@5.6.0/dist/xstate.esm.js)

---
> why?
* `npm:xstate@5.6.0` - also works (except case below)
* __deno-deploy + fresh + tailwindcss__ => on `build` step have `Directory not empty (os error 39):` ... (to deep for me to understand)
* [https://deno.land/x/xstate@xstate%405.6.0](https://deno.land/x/xstate@xstate%405.6.0) error and also when I try to open these files on denoland - access error .. (suppose it's relates to discussion above)

**Reply by kindkitchen** [reply]
my `deno.json` for `xstate`:
```json
{
  "imports": {
    "xstate": "https://esm.sh/xstate@latest?types=npm:xstate"
  }
}

```
